src/main/java/lavalink/client/io/jda/JdaLavalink.java

```diff
@@ -68,20 +68,32 @@ public JdaLink getExistingLink(Guild guild) {
     }
 
     /**
-     * May produce NullPointerException if JDA provider has not been initialised
+     * Returns the JDA instance with the {@code shardId shard ID}
+     *
+     * @param shardId the ID of the shard
+     * @return the JDA instance with the specified shard ID
+     *
+     * @throws IllegalStateException if the JDA provider has not been initialised
      */
     @SuppressWarnings({"WeakerAccess", "unused"})
-    @Nullable
+    @NonNull
     public JDA getJda(int shardId) {
+        if (jdaProvider == null) throw new IllegalStateException("JDA Provider not initialised!");
         return jdaProvider.apply(shardId);
```

Comment:
This may return null

Reply:
How?

Reply:
Won't throwing there mean that it won't return anything?

Reply:
Actually you're right, how can I fix it?

Reply:
Throw an `IllegalStateException`

Reply:
Don't I already do that?

Reply:
Not for the value returned by `apply()`

Reply:
Or you mean make the method explicitly throw `IllegalStateException` by putting `throws` in there? That seems like a bad idea to me since having to try catch that method will be annoying to do

Reply:
So `apply` can still return null is what you're saying

Reply:
Yes. You're not supposed to catch the exceptions because you are supposed to have provided a jdaProvider which never returns null

Reply:
So the function isn't what's supposed to be nullable, it's JDA that's supposed to be nullable right?

Reply:
The user is ***supposed*** to provide a jdaProvider which is not null, and which does not return null.

But we don't trust the user, so we make two checks for both of those conditions. If either fail, we throw an `IllegalStateException`. It's really not complicated.

Reply:
So do I need to check if `jdaProvider.apply` returns null then?

Reply:
Yes

Reply:
Okay, incoming